lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -167,53 +191,188 @@ static std::optional<LoadDotOperand> loadDotOperand(tt::LoadOp loadOp,
+  AxisInfo::DimVectorT shapePerCTA = ttg::getShapePerCTA(ty);
+
+  unsigned elemNumBits = getElementBitWidth(src);
+  unsigned elemNumBytes = std::max(elemNumBits / 8, 1u);
+  unsigned maxMultipleBytes = valInfo->getDivisibility(order[0]);
+  unsigned maxMultiple = std::max(maxMultipleBytes / elemNumBytes, 1u);
+  unsigned maxContig =
+      std::min(valInfo->getContiguity(order[0]), shapePerCTA[order[0]]);
+  unsigned alignment = std::min(maxMultiple, maxContig);
+  unsigned currPerThread = std::min(alignment, 128 / elemNumBits);
+
+  ttg::CTALayoutAttr CTALayout = ttg::getCTALayout(ty.getEncoding());
+  return ttg::BlockedEncodingAttr::get(loadOp->getContext(), ty.getShape(), 4,
+                                       order, numWarps, threadsPerWarp,
+                                       CTALayout);
 }
 
-/// Collect loads to pipeline. Return success if we can pipeline this loop
-static void collectOpsToPipeline(scf::ForOp forOp,
-                                 SmallVectorImpl<LoadDotOperand> &ops,
-                                 bool &hasMMAV3) {
+// Create a map from load ops to their distance to the nearest dot op and the
+// final use of the load op (another load op, or a dot op).
+static llvm::MapVector<tt::LoadOp, std::pair<int, Operation *>>
+loadOpsToDistanceAndUse(scf::ForOp forOp) {
+  llvm::MapVector<tt::LoadOp, std::pair<int, Operation *>> loadOpToDistAndUse;
+  DenseSet<Operation *> seen;
+
   ModuleOp moduleOp = forOp->getParentOfType<ModuleOp>();
   ModuleAxisInfoAnalysis axisInfoAnalysis(moduleOp);
 
+  auto isCandidate = [&](tt::LoadOp loadOp) -> bool {
+    assert(!isLoadFromTensorPtr(loadOp) &&
+           "Block ptr should have been lowered before this pass.");
+    auto ptr = loadOp.getPtr();
+    unsigned vec = axisInfoAnalysis.getPtrContiguity(ptr);
+    if (auto mask = loadOp.getMask())
+      vec = std::min<unsigned>(vec, axisInfoAnalysis.getMaskAlignment(mask));
+
+    auto tensorTy = ptr.getType().dyn_cast<RankedTensorType>();
+    if (!tensorTy)
+      return false;
+    auto ty =
+        tensorTy.getElementType().cast<tt::PointerType>().getPointeeType();
+    unsigned width = vec * ty.getIntOrFloatBitWidth();
+    // We do not pipeline all loads for the following reasons:
+    // 1. On nvidia GPUs, cp.async's cp-size can only be 4, 8 and 16.
+    // 2. It's likely that pipling small loads won't offer much performance
+    //    improvement and may even hurt performance by increasing register
+    //    pressure.
+    return (width >= 32);
+  };
+
+  std::function<void(Operation * op, int, Operation *)> dfs =
+      [&](Operation *op, int distance, Operation *use) {
+        if (!seen.insert(op).second)
+          return;
+        if (auto loadOp = dyn_cast<tt::LoadOp>(op)) {
+          if (!isCandidate(loadOp))
+            return;
+          loadOpToDistAndUse[loadOp] = std::make_pair(distance, use);
+          use = op;
+          distance++;
+        }
+        for (Value operand : op->getOperands()) {
```

Comment:
Just double checking, we do not pipeline loads that are loop carried deps (i.e. manually pipelined loops) ?

Reply:
No, we don't, that's conscious choice.